/************************************************************************************
bioCC is a fast advanced correlation calculator for basics bioinformatics file formats.
It computes Signal and Pearson correlation coefficients for densities, coverages and features.
Program allows to know correlation coefficients for the whole genome, for each chromosome
separately and for predefined regions inside chromosomes:
again for the all regions and for each region separately as well.
	
bioCC is designed to treat a bunch of files at once.

Copyright (C) 2017 Fedor Naumenko (fedor.naumenko@gmail.com)
-------------------------
Last modified: 08.06.2019
-------------------------

This program is free software. It is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the	GNU General Public License for more details.
************************************************************************************/

#include "Calc.h"
#include <fstream>

using namespace std;

const string Product::Title = "bioCC";
const string Product::Version = "1.0";
//const string Product::Descr = "advanced correlation calculator for bam/bed/wig files";
const string Product::Descr = "advanced correlation calculator";

const string OutFile = Product::Title +  "_out.txt";
const string HelpOutFile = "duplicate standard output to " + OutFile + " file";
const string InFiles = "input files";

enum eOptGroup	{ oINPUT, oTREAT, oTREAT_R, oOUTPUT, oOTHER };	// oOTHER should be the last 

const char* Options::OptGroups [] = {
	"Input", "Processing", "Region processing", "Output", "Other"
};
const BYTE Options::GroupCount = sizeof(Options::OptGroups)/sizeof(char*);

// --cc option: correlation coefficient notations
const char* CCs [] = { "P", "S" };			// corresponds to CCkey::eCC
// --total option: total coefficients notations
const char* prCCs [] = { "IND", "TOT" };	// corresponds to eTotal; totalOFF is hidden
// --sort option: sorting type notations
const char* sorts [] = { "RGN", "CC" };		// corresponds to eRS; rsOFF is hidden
// --info option: types of info notations
const char* infos [] = { "LAC", "NM", "CNT", "STAT" };	// corresponds to eInfo; iNONE is hidden

const char* ForAligns = "For the alignments only";
const char* IgnoreBed = "Ignored for the ordinary beds";

//	{ char,	str,	Signs,	type,	group,	defVal,	minVal,	maxVal,	strVal,	descr, addDescr }
// field 7: vUNDEF if value is prohibited
// field 6: vUNDEF if no default value should be printed
Options::Option Options::List [] = {
	{ 'a', "align",		0,	tENUM,	oINPUT,	FALSE,	vUNDEF, 2, NULL,
	"input bed files are alignments. Ignored for bam and wig", NULL },
	{ 'g', "gen",		0,	tNAME,	oINPUT, vUNDEF, 0, 0, NULL,
	"chromosome sizes file", NULL },
	{ HPH, "gap-len",	8,	tINT,	oINPUT,	1000, 50, 1e5, NULL,
	"minimal length of undefined nucleotide region in genome\nwhich is declared as a gap.\nIgnored for the chromosome sizes file and for the ordinary beds", NULL },
	{ 'd', "dupl",		8,	tENUM,	oINPUT, TRUE,	0, 2, (char*)Options::Booleans,
	"accept duplicate reads.", ForAligns },
	//{ HPH, "diff-sz",	0,	tENUM,	oINPUT, FALSE,	0, 2, (char*)Options::Booleans,
	//"allow to ignore reads with different size.", ForAligns },
	{ 'l', "list",		0,	tNAME,	oINPUT, vUNDEF, 0, 0, NULL,
	"list of multiple input files.\nFirst (primary) file in list is comparing with others (secondary)", NULL },
	{ 'c', Chrom::Abbr,	0,	tNAME,	oTREAT, vUNDEF, 0, 0, NULL,	"treat specified chromosome only", NULL },
	{ 'r', "cc",		0,	tCOMB,	oTREAT,	CCkey::ccP, CCkey::ccP, CCkey::ccS, (char*)CCs,
	"correlation coefficient, in any combination: ? - Pearson, ? - signal", NULL },
	{ 's', "space",		0,	tINT,	oTREAT,	200, 2, 1e4, NULL,
	"resolution: span in bp by which reads will be counted\nto define a density.", ForAligns },
	{ 'f', "fbed",		0,	tNAME,	oTREAT_R, vUNDEF,	0, 0, NULL,
	"'template' ordinary bed file which features define compared regions.\n", IgnoreBed},
	{ 'e', "ext-len",	0,	tINT,	oTREAT_R,0, 0, 1e4, NULL,
	"length by which the features in primary file (for ordinary beds) or in\n'template' (for alignments and wigs) will be extended in both directions\nbefore treatment", NULL },
	{ HPH, "ext-step",	0,	tINT,	oTREAT_R,0, 0, 500, NULL,
	"step of extending features in primary bed file;\nif 0 then no step calculation. For the ordinary beds only", NULL },
	{ 'C',	"pr-cc",	0,	tCOMB,	oOUTPUT, Results::cIND, Results::cIND, Results::cTTL, (char*)prCCs,
	"print coefficient, in any combination:\n? - for each chromosome individually, ? - total", NULL },
	{ 'B', "bin-width",	0,	tFLOAT,	oOUTPUT,0, 0, 1.0F, NULL,
	"print frequency histogram with given bin width", NULL },
	{ 'S', "sort",	0,	tENUM,	oOUTPUT, rsOFF,	rsR, rsC, (char*)sorts,
	"print region coefficients, sorted by:\n? - regions, ? - coefficients", NULL },
	{ HPH, "norm",	0,	tENUM,	oTREAT_R,TRUE,	0, 2, (char*)Options::Booleans,
	"normalize regions before calculation.", IgnoreBed },
	//{ HPH, "cross",	0, tBOOL, FALSE, 0, 0, NULL, "cross-correlation" },
	//{ HPH, "circ",	0, tBOOL, FALSE, 0, 0, NULL, "circular cross-correlation" },
	//{ 's', "step",	0, tINT, 2e5, 1, (float)INT_MAX, NULL, "minimal shift between outputted coefficient during cross-correlation." },
	{ 'i', "info",	0,	tENUM, oOUTPUT,	Obj::iNM, Obj::iLAC, Obj::iSTAT, (char*)infos,
	"print information about file:\n? - laconic, ? - name only, ? - number of items, ? - statistics", NULL },
	{ 'w', "warn",	8,	tENUM, oOUTPUT,	FALSE,	vUNDEF, 2, NULL,
	"print each file's item ambiguity, if they exist.", NULL },
	{ 'o', "out",	0,	tENUM,	oOUTPUT,FALSE,	vUNDEF, 2, NULL, HelpOutFile.c_str(), NULL },
	{ 't', "time",	0,	tENUM,	oOTHER,	FALSE,	vUNDEF, 2, NULL, "print run time", NULL },
	{ 'v', Version,	0,	tVERS,	oOTHER,	vUNDEF, vUNDEF, 0, NULL, "print program's version", NULL },
	{ HPH, "summ",	8,	tSUMM,	oOTHER,	vUNDEF, vUNDEF, 0, NULL, "print program's summary", NULL },
	{ 'h', "help",	0,	tHELP,	oOTHER,	vUNDEF, vUNDEF, 0, NULL, "print usage information", NULL }
};
const BYTE	Options::OptCount = sizeof(Options::List)/sizeof(Options::Option);

const Options::Usage Options::Usages[] = {	// content of 'Usage' variants in help
	{ vUNDEF, "<file1> <file2> ...", true, NULL },
	{ oFILE_LIST, NULL, true, NULL }
};
const BYTE Options::UsageCount = sizeof(Options::Usages)/sizeof(Options::Usage);

dostream dout;	// stream's duplicator

/*****************************************/
int main(int argc, char* argv[])
{
	int fileInd = Options::Parse(argc, argv);
	if( fileInd < 0 )	return 1;		// wrong option or tip output
	int ret = 0;						// main() return code
	const FileList* fList = NULL;
	
	Chrom::SetCustomOption(oCHROM);
	if(Options::GetBVal(oOUTFILE))	dout.OpenFile(OutFile);
	Timer::Enabled = Options::GetBVal(oTIME);
	Timer timer;
	try {
		char **inFiles;
		short cntInFiles;
		const char* gName = Options::GetSVal(oGEN);		// genome

		// check file-list is set & exist
		const char* fListName = Options::GetSVal(oFILE_LIST);
		if( fListName ) {
			fList = new FileList( FS::CheckedFileName(fListName) );
			inFiles = fList->Files();
			cntInFiles = fList->Count();
			if( cntInFiles == 0 )
				Err(Err::MISSED, NULL, InFiles + " (no uncommented line in " + 
					string(fListName) + ")").Throw();
		}
		else	inFiles = argv + fileInd, cntInFiles = argc - fileInd;
		if( cntInFiles < 2 )			// check input files
			Err(Err::MISSED, NULL, cntInFiles ? "secondary " + InFiles : InFiles).Throw();

		if(!gName && !FS::HasExt(inFiles[0], FT::Ext(FT::BAM)))
			Err(Options::OptionToStr(oGEN) + " is required for all the file types except BAM",
			inFiles[0]).Throw();

		ChromSizes cSizes(gName, NULL, true);	
		DefRegions gRgn(cSizes, Options::GetIVal(oGAPLEN));

		CorrPair cPair(inFiles[0], gRgn, Options::GetSVal(oFBED), cntInFiles > 2);
		for(short i=1; i<cntInFiles; i++)
			cPair.CalcCC(inFiles[i]);
	}
	catch(const Err &e)			{ ret = 1;	cerr << e.what() << EOL; }
	catch(const exception &e)	{ ret = 1;	cerr << e.what() << EOL; }
	if(fList)	delete fList;
	timer.Stop(false);
	return ret;
}


/************************ class FileList ************************/
#ifdef OS_Windows
// Returns true if 'name' is file's pattern
bool	IsFilePattern	(const char* name)
{
	return strchr(name, '*') != NULL || strchr(name, '?') != NULL;
}

string GetPath	(const LPCTSTR name)
{
	const char* pch = strrchr(name, '/');
	return pch ? string(name, pch-name+1) : strEmpty;
}

// Fills vector by filenames according name template.
// Works only if _UNICODE is not defined.
//	@files: vector of file names
//	@templ: name template, which can include '*' and '?' marks
void FillFilesByTemplate(vector<string>& files, const LPCTSTR templ)
{
	string path = GetPath(templ);
	WIN32_FIND_DATA ffd;
 
	// check directory and estimate listFileNames capacity
	HANDLE hFind = FindFirstFile( templ, &ffd );
	if( hFind == INVALID_HANDLE_VALUE )		
		Err("bad file or content", templ).Throw();
	if( files.capacity() == 0 ) {		// count files to reserve capacity
		short count = 1;
		for(; FindNextFile(hFind, &ffd); count++);
		files.reserve(count);
		hFind = FindFirstFile( templ, &ffd );
	}
	// fill the list
	do	files.push_back(path + string(ffd.cFileName));	//  works only if _UNICODE isn't defined
	while (FindNextFile(hFind, &ffd));
	FindClose(hFind);
}
#endif	// OS_Windows

FileList::FileList(char* files[], short cntFiles) : _files(NULL), _memRelease(true)
{
#ifdef OS_Windows
	short i;
	bool hasTemplate;
	for(i=0; i<cntFiles; i++)
		if( hasTemplate = IsFilePattern(files[i]) )
			break;
	if( hasTemplate ) {
		// First we fill vector of file names, second initialize _files by this vector.
		// It needs because files[] may contain file names and file template (pattern) as well,
		// so in case of Windows we don't know the total amount of files beforehand.
		vector<string> tmpFiles;
		if( cntFiles > 1 )
			tmpFiles.reserve(cntFiles);
		// else if it's a single template name, capacity will be reserved in FillFilesByTemplate(),
		// or if it's a single common name, capacity will not be reserved at all

		for(i=0; i<cntFiles; i++)
			if( IsFilePattern(files[i]) )
				FillFilesByTemplate(tmpFiles, files[i]);
			else		// real list of files
				tmpFiles.push_back(files[i]);

		_files = new char*[_count=tmpFiles.size()];
		for(i=0; i<_count; i++) {
			_files[i] = (char*)malloc(tmpFiles[i].length()+1);
			strcpy(_files[i], tmpFiles[i].c_str());
		}
	}
	else 
#endif	// OS_Windows
	{
		_files = files;
		_count = cntFiles;
		_memRelease = false;
	}
}

FileList::FileList(const char* fName) : _files(NULL), _memRelease(true)
{
	TabFile file(fName);
	ULONG cntLines;
	char *dstStr;
	const char *srcStr;
	// no needs to check since aborting invalid file is set
	const char *currLine = file.GetFirstLine(&cntLines);
	vector<char*> tmpFiles;	// temporary vector because cntLines is not proof, but estimated capacity
	
	_count = 0;
	tmpFiles.reserve(cntLines);
	while(currLine!=NULL) {
		srcStr = file.StrField(0);
		dstStr = (char*)malloc(strlen(srcStr)+1);
		strcpy(dstStr, srcStr);
		tmpFiles.push_back(dstStr);
		_count++;
		currLine=file.GetLine();
	}
	_files = new char*[_count];
	for(short i=0; i<_count; i++)
		_files[i] = tmpFiles[i];
}

/************************ end of class FileList ************************/

